src/main/java/io/debezium/connector/jdbc/JdbcChangeEventSink.java

```diff
@@ -40,71 +42,181 @@ public class JdbcChangeEventSink implements ChangeEventSink {
+            if (optionalTableId.isEmpty()) {
+
+                LOGGER.warn("Ignored to write record from topic '{}' partition '{}' offset '{}'. No resolvable table name", record.topic(), record.kafkaPartition(),
+                        record.kafkaOffset());
+                continue;
+            }
+
+            SinkRecordDescriptor sinkRecordDescriptor = buildRecordSinkDescriptor(record);
+
+            final TableId tableId = optionalTableId.get();
+
+            if (sinkRecordDescriptor.isTombstone()) {
+                LOGGER.debug("Skipping tombstone record {}", sinkRecordDescriptor);
+                continue;
+            }
+
+            if (sinkRecordDescriptor.isTruncate()) {
+
+                if (!config.isTruncateEnabled()) {
+                    LOGGER.debug("Truncates are not enabled, skipping truncate for topic '{}'", sinkRecordDescriptor.getTopicName());
+                    continue;
+                }
+
+                // Here we want to flush the buffer to let truncate having effect on the buffered events.
+                flushBuffers(updateBufferByTable);
+
+                flushBuffers(deleteBufferByTable);
+
+                try {
+                    final TableDescriptor table = checkAndApplyTableChangesIfNeeded(tableId, sinkRecordDescriptor);
+                    writeTruncate(dialect.getTruncateStatement(table));
+                }
+                catch (SQLException e) {
+                    throw new ConnectException("Failed to process a sink record", e);
+                }
+            }
+
+            if (sinkRecordDescriptor.isDelete()) {
+
+                if (!config.isDeleteEnabled()) {
+                    LOGGER.debug("Deletes are not enabled, skipping delete for topic '{}'", sinkRecordDescriptor.getTopicName());
+                    continue;
+                }
+
+                RecordBuffer tableIdBuffer = deleteBufferByTable.computeIfAbsent(tableId, k -> new RecordBuffer(config));
+                List<SinkRecordDescriptor> toFlush = tableIdBuffer.add(sinkRecordDescriptor);
+
+                flushBuffer(tableId, toFlush);
+            }
+            else {
+
+                if (deleteBufferByTable.get(tableId) != null && !deleteBufferByTable.get(tableId).isEmpty()) {
+                    // When an insert arrives, delete buffer must be flushed to avoid losing an insert for the same record after its deletion.
+                    // this because at the end we will always flush inserts before deletes.
+
+                    flushBuffer(tableId, deleteBufferByTable.get(tableId).flush());
+                }
+
+                RecordBuffer tableIdBuffer = updateBufferByTable.computeIfAbsent(tableId, k -> new RecordBuffer(config));
+
+                List<SinkRecordDescriptor> toFlush = tableIdBuffer.add(sinkRecordDescriptor);
+                flushBuffer(tableId, toFlush);
+            }
+
+        }
+
+        flushBuffers(updateBufferByTable);
```

Comment:
TBH I don't think we should do flush at the end of the method as it means the buffering granularity is only per batch.
I'd propose to

* remove the calls at the end of the method
* add calls to the `close` method
* trigger the flush when there are no records coming for certain amount of time - probably configurable to provide trade-off between batching/latrency

Reply:
I have started the discussion about this approach some weeks ago [here](https://debezium.zulipchat.com/#narrow/stream/383534-community-jdbc/topic/Batching/near/397516639). @Naros pointed the problem on how to manage the processed offset in this case since you cannot update it until you effectively flush. It think this requires more thinking and maybe can be a future enhancement. 

Also is there any real case where you want to have `consumer.max.poll.records` < `batch.size`?

[Confluent JDBC Connector](https://docs.confluent.io/kafka-connectors/jdbc/current/sink-connector/sink_config_options.html#sink-config-options) uses the same approach.  

WDYT?

Reply:
@mfvitale Could you please create a follow-up Jira for it? IMHO the solution is pretty simple and is mirror of `DatabaseHeartbeatImpl`. We would intorduce a config option for heartbeat. A heartbeat thread will be started that will periodically write messages to a Kafka topic. This will force Kafka Connect to call our connector task and flush buffers if there is no traffic. WDYT?

Reply:
Follow-up issue created https://issues.redhat.com/browse/DBZ-7122

Is there any difference between heartbeat and a scheduled thread approach?

Reply:
I think Connect is the one who manages offsets so it might be "cleaner" to go using "official" path.